state/machine_linklayerdevices.go

```diff
@@ -1120,3 +1121,115 @@ func generateMACAddress() string {
+func addAddressToResult(networkInfos []network.NetworkInfo, address *Address) ([]network.NetworkInfo, error) {
+	ifaceAddress := network.InterfaceAddress{
+		Address: address.Value(),
+		CIDR:    address.SubnetCIDR(),
+	}
+	for i := range networkInfos {
+		networkInfo := &networkInfos[i]
+		if networkInfo.InterfaceName == address.DeviceName() {
+			networkInfo.Addresses = append(networkInfo.Addresses, ifaceAddress)
+			return networkInfos, nil
+		}
+	}
+
+	MAC := ""
+	device, err := address.Device()
+	if err == nil {
+		MAC = device.MACAddress()
+	} else if !errors.IsNotFound(err) {
+		return nil, err
+	}
+	networkInfo := network.NetworkInfo{
+		InterfaceName: address.DeviceName(),
+		MACAddress:    MAC,
+		Addresses:     []network.InterfaceAddress{ifaceAddress},
+	}
+	return append(networkInfos, networkInfo), nil
+}
+
+// GetNetworkInfoForSpaces returns MachineNetworkInfoResult with a list of devices for each space in spaces
+// TODO(wpk): 2017-05-04 This does not work for L2-only devices as it iterates over addresses, needs to be fixed.
+// When changing the method we have to keep the ordering.
+func (m *Machine) GetNetworkInfoForSpaces(spaces set.Strings) map[string](MachineNetworkInfoResult) {
+	results := make(map[string](MachineNetworkInfoResult))
+
+	var privateAddress network.Address
+
+	if spaces.Contains("") {
+		var err error
+		privateAddress, err = m.PrivateAddress()
+		if err != nil {
+			error := errors.Annotatef(err, "getting machine %q preferred private address", m.MachineTag())
+			results[""] = MachineNetworkInfoResult{Error: &error}
+			spaces.Remove("")
+		}
+	}
+
+	addresses, err := m.AllAddresses()
+	logger.Debugf("Looking for something from spaces %v in %v", spaces, addresses)
+	if err != nil {
+		newErr := errors.Annotate(err, "cannot get devices addresses")
+		result := MachineNetworkInfoResult{Error: &newErr}
+		for space := range spaces {
+			if _, ok := results[space]; !ok {
+				results[space] = result
+			}
+		}
+		return results
+	}
+	actualSpaces := set.NewStrings()
+	for _, addr := range addresses {
+		subnet, err := addr.Subnet()
+		switch {
+		case errors.IsNotFound(err):
+			logger.Debugf("skipping %s: not linked to a known subnet (%v)", addr, err)
+		case err != nil:
+			logger.Errorf("cannot get subnet for address %q - %q", addr, err)
+		default:
+			space := subnet.SpaceName()
+			actualSpaces.Add(space)
+			if spaces.Contains(space) {
+				r := results[space]
+				r.NetworkInfos, err = addAddressToResult(r.NetworkInfos, addr)
+				if err != nil {
+					r.Error = &err
+				} else {
+					results[space] = r
+				}
+			}
+			if spaces.Contains("") && privateAddress.Value == addr.Value() {
+				r := results[""]
+				r.NetworkInfos, err = addAddressToResult(r.NetworkInfos, addr)
+				if err != nil {
+					r.Error = &err
+				} else {
+					results[""] = r
+				}
+			}
+		}
+	}
+
+	actualSpacesStr := strings.Join(actualSpaces.Values(), ",")
```

Comment:
there is network.QuoteSpaces which produces output that is a bit nicer when you have a list of possible spaces which might contain and empty string space name.